### PR TITLE
Cortex-R: Use .macro for common assembly code

### PR DESCRIPTION
**Pull request by bbolen**
A couple of cleanups for Cortex-R.  Use a define for the svc call instead of a hardcoded number.  Factor out some exception code into an assembly macro to make exception handling clearer.

### CODE REVIEW

File: arch/arm/core/aarch32/cortex_a_r/exc.S [hunk 1]
```diff
@@ -38,30 +38,12 @@ GTEXT(z_arm_undef_instruction)
...
-	 * mode is Thumb; otherwise, it is offset by 4.
-	 */
-	push {r0}
-	mrs r0, spsr
-	tst r0, #T_BIT
-	subeq lr, #4	/* ARM   (!T_BIT) */
-	subne lr, #2	/* Thumb (T_BIT) */
-	pop {r0}
-
+.macro exception_entry, mode
```
Comment (stephanosio) [inline arch/arm/core/aarch32/cortex_a_r/exc.S:41]: IIRC, GNU assembler macro syntax is supposed to be something like `.macro [name] [param1], [param2], ...`, so there shouldn't be a comma after the macro name.

File: arch/arm/core/aarch32/cortex_a_r/exc.S [hunk 2]
```diff
@@ -196,58 +154,9 @@ SECTION_SUBSEC_FUNC(TEXT, __exc, z_arm_data_abort)
...
-
-	cps #MODE_SYS
-	stm r1, {r4-r11, sp}
-	cps #MODE_ABT
-#endif
-
-	/* Invoke fault handler */
-	mov r0, sp
+	exception_entry MODE_ABT
 	bl z_arm_fault_data
```
Comment (stephanosio) [inline arch/arm/core/aarch32/cortex_a_r/exc.S:159]: Spurious empty line after `bl` (it would be nice to keep it consistent with the rest).

**Review by stephanosio — COMMENTED** [review]
I am quite sure the exception entry and exit code will become different for each exception type as more advanced handling features are added (which, by the way, was why I initially did not bother refactoring them); but, there is currently no plan to implement such advanced features in the near future and, noting that, this refactoring looks reasonable.

Other than the questionable macro definition syntax used here, it looks ok to me.

File: arch/arm/core/aarch32/cortex_a_r/exc.S (hunk 1 above)
Comment (bbolen) [reply arch/arm/core/aarch32/cortex_a_r/exc.S]: fixed

File: arch/arm/core/aarch32/cortex_a_r/exc.S (hunk 2 above)
Comment (bbolen) [reply arch/arm/core/aarch32/cortex_a_r/exc.S]: fixed

**Review by stephanosio — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Comment by ioannisg** [comment]
Maybe enhance the PR title a bit, @bbolen ?

**Pull request merged**